### PR TITLE
IAP + CDN e2e testing implementation

### PR DESCRIPTION
**Pull request by rramkumar1**
Note that this PR also contains some small changes to the validator:
    1.  In order to prevent the http client in the validator from following redirects, I implemented the 
         CheckRedirect hook. 
    2.  Move the pathToDefaultBackend variable up higher so that other files can use it.

Ref: #301

This change is [Reviewable](https://reviewable.io/reviews/kubernetes/ingress-gce/319)

### CODE REVIEW

File: pkg/fuzz/features/cdn.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import (
+	"fmt"
+	"net/http"
+
+	"k8s.io/api/extensions/v1beta1"
+	"k8s.io/ingress-gce/pkg/fuzz"
+)
+
+// Cdn is a feature in BackendConfig that supports using GCP CDN.
+var Cdn = &CDNFeature{}
```
Comment (bowei) [inline pkg/fuzz/features/cdn.go]: cdn? (no need to export?)

File: pkg/fuzz/features/cdn.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+func (v *cdnValidator) ConfigureAttributes(env fuzz.ValidatorEnv, ing *v1beta1.Ingress, a *fuzz.IngressValidatorAttributes) error {
+	// Capture the env for use later in CheckResponse.
+	v.ing = ing
+	v.env = env
+	return nil
+}
+
+// ModifyRequest implements fuzz.FeatureValidator.
+func (v *cdnValidator) ModifyRequest(host, path string, req *http.Request) {
+	// Warm up the cache, regardless of whether this host + path has CDN enabled.
```
Comment (bowei) [inline pkg/fuzz/features/cdn.go]: should probably comment that // cache=true makes the echo server add caching headers

File: pkg/fuzz/features/features.go [hunk 3]
```diff
@@ -25,4 +25,6 @@ import "k8s.io/ingress-gce/pkg/fuzz"
 var All = []fuzz.Feature{
 	AllowHTTP,
 	PresharedCert,
+	Cdn,
```
Comment (bowei) [inline pkg/fuzz/features/features.go]: CDN

File: pkg/fuzz/features/features.go [hunk 4]
```diff
@@ -25,4 +25,6 @@ import "k8s.io/ingress-gce/pkg/fuzz"
...
+	Iap,
```
Comment (bowei) [inline pkg/fuzz/features/features.go]: IAP (golang code style)

File: pkg/fuzz/features/iap.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+// CheckResponse implements fuzz.FeatureValidator.
+func (v *iapValidator) CheckResponse(host, path string, resp *http.Response, body []byte) (fuzz.CheckResponseAction, error) {
+	backendConfig, err := fuzz.BackendConfigForPath(host, path, v.ing, v.env)
+	if err != nil {
+		return fuzz.CheckResponseContinue, err
+	}
+	var iapEnabled bool
+	if backendConfig.Spec.Iap != nil && backendConfig.Spec.Iap.Enabled == true {
+		iapEnabled = true
+	}
```
Comment (bowei) [inline pkg/fuzz/features/iap.go]: do you need iapEnabled? why not return if false?

File: pkg/fuzz/features/iap.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+			glog.V(2).Infof("The response was %v", resp)
+			return fuzz.CheckResponseContinue, fmt.Errorf("IAP is turned on but response %v did not return a 302", resp)
+		}
+		if resp.Header.Get("x-goog-iap-generated-response") == "" {
+			return fuzz.CheckResponseContinue, fmt.Errorf("IAP is turned on but response w/ header %v did not contain IAP header", resp.Header)
+		}
+		return fuzz.CheckResponseSkip, nil
+	}
+	// If IAP is turned off, verify that the response code was not 302.
+	if !iapEnabled && resp.StatusCode == http.StatusFound {
```
Comment (bowei) [inline pkg/fuzz/features/iap.go]: don't need this?, standard check will be for 200

File: pkg/fuzz/features/cdn.go (hunk 1 above)
Comment (rramkumar1) [reply pkg/fuzz/features/cdn.go]: Since I made this fully capitalized now, it will technically be exported.

File: pkg/fuzz/features/cdn.go (hunk 2 above)
Comment (rramkumar1) [reply pkg/fuzz/features/cdn.go]: Done..

File: pkg/fuzz/features/features.go (hunk 3 above)
Comment (rramkumar1) [reply pkg/fuzz/features/features.go]: Done.

File: pkg/fuzz/features/features.go (hunk 4 above)
Comment (rramkumar1) [reply pkg/fuzz/features/features.go]: Done.

File: pkg/fuzz/features/iap.go (hunk 5 above)
Comment (rramkumar1) [reply pkg/fuzz/features/iap.go]: See comment below.

File: pkg/fuzz/features/iap.go (hunk 6 above)
Comment (rramkumar1) [reply pkg/fuzz/features/iap.go]: The main reason was to return a custom error.

**Comment by bowei** [comment]
We need to tag the image used for guitar tests. Right now as it targets master, the tests will start failing as it is testing the old ingress controller. Let's talk tomorrow.

/lgtm

File: pkg/fuzz/helpers.go [hunk 7]
```diff
@@ -17,13 +17,59 @@ limitations under the License.
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/ingress-gce/pkg/annotations"
 	backendconfig "k8s.io/ingress-gce/pkg/apis/backendconfig/v1beta1"
+	backendconfigutil "k8s.io/ingress-gce/pkg/backendconfig"
+	translatorutil "k8s.io/ingress-gce/pkg/controller/translator"
 )
 
+// BackendConfigForPath returns the BackendConfig associated with the given path.
+// Note: This function returns an empty object (not nil pointer) if a BackendConfig
```
Comment (bowei) [inline pkg/fuzz/helpers.go]: we probably want to distinguish these two cases for testing BackendConfig does not exist vs BackendConfig exists but is empty.

Comment (rramkumar1) [reply pkg/fuzz/helpers.go]: You mean have separate function for each case? I can take care of it in a separate PR.

**Pull request merged**